### PR TITLE
Improve error handling for bad continuation tokens and search values.

### PR DESCRIPTION
**Pull request by johnstairs**
## Description
Improves handling for malformed SQL and Cosmos DB continuation tokens, and improves error handling when parsing search values. Also avoids being over broad with our error handling. For instance, `ArgumentException`s, like those raised in `EnsureArg` calls, would be translated to 400 response codes instead of 500.

## Related issues
Addresses #1259, #674.
Addresses part of #1342.

## Testing
Added E2E test cases.

### CODE REVIEW

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs [hunk 1]
```diff
@@ -278,7 +279,15 @@ internal async Task<FeedResponse<T>> ExecuteDocumentQueryAsync<T>(QueryDefinitio
 
             var documentQuery = _cosmosQueryFactory.Create<T>(_containerScope.Value, context);
 
-            return await documentQuery.ExecuteNextAsync(cancellationToken);
+            try
+            {
+                return await documentQuery.ExecuteNextAsync(cancellationToken);
+            }
+            catch (CosmosException e) when (e.StatusCode == HttpStatusCode.BadRequest && continuationToken != null && e.ResponseBody.StartsWith("Malformed continuation token", StringComparison.OrdinalIgnoreCase))
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs:286]: Did this pass through https://github.com/microsoft/fhir-server/blob/master/src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs#L45? Should this be in the handler?

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs:286]: Makes sense. I was thinking that we should only return a 400 if this was a CT that came from a client, not one that we used internally. But that is probably unlikely.

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs:286]: @brandonpollett, it does not pass though the handler:
image
I don't know if this is a bug in the SDK or our code.

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs:286]: Thanks, I'll make a bug to look into this

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs:286]: #1390

**Review by brendankowitz — APPROVED** [review]

**Pull request merged**